### PR TITLE
Applicant can edit their Membership Application

### PR DESCRIPTION
**Pull request by tochman**
PT Story: https://www.pivotaltracker.com/story/show/134078325

_As an applicant
In order to be able to edit my application
I want to be allowed to do that_

Changes proposed in this pull request:
1. adds User model with Devise for authentication
2. adds relation User -> MembershipApplication
3. adds Edit view for MembershipApplication
4. extracts form fields to partial
5. introduces authorization engine (Pundit)

Ready for review @thesuss @weedySeaDragon @diraulo et al

### CODE REVIEW

File: features/show_membership_application.feature [hunk 1]
```diff
@@ -3,21 +3,27 @@ Feature: As an Admin
   I need to review a Membership Application that has been submitted
   PT: https://www.pivotaltracker.com/story/show/133950343
 
-Background:
-Given the following applications exist:
-  | company_name | company_number | contact_person | phone_number | company_email |
-  | Hunderiet    | 1234567890     | Emma Svensson  | 1234-234567  | min@mail.se   |
-  | DoggieZone   | 2345678901     | Pam Andersson  | 0234-234567  | din@mail.se   |
-  | Tassa-in AB  | 1234367890     | Anna Knutsson  | 1234-234569  | sin@mail.se   |
+  Background:
```
Comment (weedySeaDragon) [inline features/show_membership_application.feature:6]: This hasn't been implemented yet.  So don't we first need a feature that does this?

**Comment by weedySeaDragon** [comment]
Where does the visitor log in so they become the `current_user`?  

This is the part I haven't been clear about:  This PR seems to assume that a visitor has received their login credentials (user id & password), and that there are the UI pieces in place so that a visitor can log in.  Specifically, the Background `Given the following users exists ...`   has not yet been implemented.  

Don't we need a feature (or features) for this? _(A visitor receives their login credentials.  A visitor can log in)_

**Comment by thesuss** [comment]
It needs to be tested too, yes. And prettified. But both could be done in a separate feature. I agree that it's missing, but it's also not critical at the moment. 

That's how I have understood it.

**Comment by weedySeaDragon** [comment]
When I run the feature tests, they pass because the `Background` steps have been automagically made to happen.  But if I try to do those same steps manually, I can't.  _There's no way to do them manually._

I understand _why_ you would want to separate the testing such that you have the background steps automagically take care of separate functionality. (= Keeping this functionality/PR well bounded)   But at some point, somewhere in the system the user has to actually be able to do those things: _get log in credentials_ and _log in_

I'm either missing something about how the users would use the system, or I'm not following something about how this PR/user story relates to other user stories in Pivotal Tracker -- or both.  

(And if it's better to move this discussion somewhere else, just let me know.)

**Comment by tochman** [comment]
That is correct. There is no way to test this in the browser/manually at the moment. Well, there is, but it is not easy. There are no links to Login in or Log out or Register pages. But the routes are there. Run `rails routes` to see what url's are part of the system. You can then enter those paths manually in the browser window.

File: features/show_membership_application.feature (hunk 1 above)
Comment (tochman) [reply features/show_membership_application.feature:6]: It was implemented in https://github.com/AgileVentures/shf-project/pull/13

Comment (weedySeaDragon) [reply features/show_membership_application.feature:6]: `"Given the following user exists..."`  hasn't been implemented yet.  #13 was about the membership application, not users & devise.

But given your comment in the overall conversation about this,  I understand now.  (or at least I'm less confused :-) )

**Comment by weedySeaDragon** [comment]
I think I'm getting it now.  :-)  

My confusions stems from the fact that:
1.  There are no explicit features in Pivotal Tracker for doing the UI for a user logging in, nor is there anything about how the user will get their credentials. I think we all understand that must happen, but it hasn't been made explicit.    I'm not yet clear on why those aren't explicit features in Pivotal Tracker, but you all are clearly good with how things are.  So I'll wait and watch and learn :-) .
2. I'm used to bootstrapping an app by having these typical 'login, get credentials...' done quite early.  The client reviews the basic _getting credentials & logging in_ while the first domain functionality is being done. (It's more of a project management decision to do it that way.)  That's _not_ a 'better' way; it's just the way that was stuck in my head, so I had a hard time seeing _this_ way.

File: app/views/partials/_top.html.haml [hunk 2]
```diff
@@ -21,7 +21,10 @@
             %li.page_item.page-item-2
               = link_to 'Mer', 'http://sverigeshundforetagare.se/'
             %li.page_item.page-item-2
-              = link_to 'Apply for membership', new_membership_path
+              - if current_user.try(:membership_applications).try(:any?) && current_user.try(:membership_applications).try(:last).try(:persisted?)
```
[the remark this answers is not in the document]
Comment (tochman) [reply app/views/partials/_top.html.haml:24]: This line is up for debate. This is a temp solution. We would probably want to use Pundit or some helper to check for this. I would argue that this is a refactoring ticket candidate. But I don't know.

[the remark this answers is not in the document]
Comment (thesuss) [reply app/views/partials/_top.html.haml:24]: Lets have it like that for now. We can ignore the hound barks until we have figured out a better setup, and then make a PR with that change.

**Review by thesuss — APPROVED** [review]

File: app/views/partials/_top.html.haml (hunk 2 above)
[the remark this answers is not in the document]
Comment (tochman) [reply app/views/partials/_top.html.haml:24]: I'm okay with that approach

File: features/create_membership_application.feature [hunk 3]
```diff
@@ -4,6 +4,12 @@ Feature: As a visitor
 
   PT Feature: https://www.pivotaltracker.com/story/show/133940725
```
Comment (thesuss) [inline features/create_membership_application.feature]: Could you please make this on line four and just PT: link ? 
The second instance of the word "Feature" gets coloured in some IDE's and that could be confusing.

File: features/edit_membership_application.feature [hunk 4]
```diff
@@ -0,0 +1,31 @@
+Feature: As an applicant
+  In order to be able to edit my application
+  I want to be allowed to do that
+
+  PT Feature: https://www.pivotaltracker.com/story/show/134078325
```
Comment (thesuss) [inline features/edit_membership_application.feature]: Could you please make this on line four and just PT: link ? 
The second instance of the word "Feature" gets coloured in some IDE's and that could be confusing.

**Review by thesuss — CHANGES_REQUESTED** [review]

File: features/edit_membership_application.feature (hunk 4 above)
Comment (tochman) [reply features/edit_membership_application.feature]: Okay.

**Review by thesuss — APPROVED** [review]

**Comment by tochman** [comment]
Guys, this feature wasn't complete. I was a bit stressed and forgot to update the `edit` and `update` actions as well as the scenario in `edit_membership_application.feature:19` There is no assertion in that scenario and we have no way of telling if the Company Name actually gets updated to "A Doggy Dog World" (it doesn't).

Lessons learned: 
1. Don't code with small children present
2. Don't trust that your teammates actually read through the acceptance tests when doing code review ;--)

Action points for @thesuss : Make this a part of the ongoing work with setting the status of an application.

**Pull request merged**